modules/renter/proto/merkleroots.go

```diff
@@ -0,0 +1,375 @@
+}
+
+// moveLastCachedSubTreeToUncached deletes the last cached subTree and appends
+// its elements to the uncached roots.
+func (mr *merkleRoots) moveLastCachedSubTreeToUncached() error {
+	mr.cachedSubTrees = mr.cachedSubTrees[:len(mr.cachedSubTrees)-1]
+	rootIndex := len(mr.cachedSubTrees) * merkleRootsPerCache
+	roots, err := mr.merkleRootsFromIndex(rootIndex, mr.numMerkleRoots)
+	if err != nil {
+		return errors.AddContext(err, "failed to read cached tree's roots")
+	}
+	mr.uncachedRoots = append(mr.uncachedRoots, roots...)
+	return nil
+}
+
+// push appends a merkle root to the end of the contract. If the number of
+// uncached merkle roots grows too big we cache them in a new subTree.
+func (mr *merkleRoots) push(root crypto.Hash) error {
+	// Sanity check the number of uncached roots before adding a new one.
+	if len(mr.uncachedRoots) == merkleRootsPerCache {
+		build.Critical("the number of uncachedRoots is too big. They should've been cached by now")
+	}
+	// Calculate the root offset within the file and write it to disk.
+	rootOffset := fileOffsetFromRootIndex(mr.len())
+	if _, err := mr.file.WriteAt(root[:], rootOffset); err != nil {
+		return err
+	}
+	// Add the root to the unached roots.
+	mr.appendRoot(root)
+
+	// Increment the number of roots.
+	mr.numMerkleRoots++
+	return nil
+}
+
+// root returns the root of the merkle roots.
+func (mr *merkleRoots) root() crypto.Hash {
+	tree := crypto.NewCachedTree(sectorHeight)
+	for _, st := range mr.cachedSubTrees {
+		if err := tree.PushSubTree(st.height, st.sum); err != nil {
+			// This should never fail.
+			build.Critical(err)
+		}
+	}
+	for _, root := range mr.uncachedRoots {
+		tree.Push(root)
+	}
+	return tree.Root()
+}
+
+// newRoot returns the root of the merkleTree after appending the newRoot
+// without actually appending it.
+func (mr *merkleRoots) newRoot(newRoot crypto.Hash) crypto.Hash {
+	tree := crypto.NewCachedTree(sectorHeight)
+	for _, st := range mr.cachedSubTrees {
+		if err := tree.PushSubTree(st.height, st.sum); err != nil {
+			// This should never fail.
+			build.Critical(err)
+		}
+	}
+	for _, root := range mr.uncachedRoots {
+		tree.Push(root)
+	}
+	// Push the new root.
+	tree.Push(newRoot)
+	return tree.Root()
+}
+
+// merkleRoots reads all the merkle roots from disk and returns them. This is
+// not very fast and should only be used for testing purposes.
+func (mr *merkleRoots) merkleRoots() (roots []crypto.Hash, err error) {
+	// Get roots.
+	roots, err = mr.merkleRootsFromIndex(0, mr.numMerkleRoots)
+	if err != nil {
+		return nil, err
+	}
+	// Sanity check: should have read exactly numMerkleRoots roots.
+	if len(roots) != mr.numMerkleRoots {
+		build.Critical(fmt.Sprintf("Number of merkle roots on disk (%v) doesn't match numMerkleRoots (%v)",
+			len(roots), mr.numMerkleRoots))
+	}
+	return
+}
+
+// merkleRootsFrom index readds all the merkle roots in range [from;to)
+func (mr *merkleRoots) merkleRootsFromIndex(from, to int) ([]crypto.Hash, error) {
```

Comment:
would probably also have it be `merkleRootsFromIndexFromDisk`. Though that's a bit of a mouthful.